src/Type/Traits/ConstantScalarTypeTrait.php

```diff
@@ -0,0 +1,53 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Traits;
+
+use PHPStan\TrinaryLogic;
+use PHPStan\Type\CompoundType;
+use PHPStan\Type\CompoundTypeHelper;
+use PHPStan\Type\Type;
+
+trait ConstantScalarTypeTrait
+{
+
+	public function accepts(Type $type): bool
+	{
+		if ($type instanceof self) {
+			return $this->value === $type->value;
+		}
+
+		if ($type instanceof CompoundType) {
+			return CompoundTypeHelper::accepts($type, $this);
+		}
+
+		return false;
+	}
+
+	public function isSuperTypeOf(Type $type): TrinaryLogic
+	{
+		if ($type instanceof self) {
+			return $this->value === $type->value ? TrinaryLogic::createYes() : TrinaryLogic::createNo();
+		}
+
+		if ($type instanceof parent) {
+			return TrinaryLogic::createMaybe();
+		}
+
+		if ($type instanceof CompoundType) {
+			return $type->isSubTypeOf($this);
+		}
+
+		return TrinaryLogic::createNo();
+	}
+
+	public function generalize(): Type
+	{
+		return new parent();
+	}
+
+	public static function __set_state(array $properties): Type
+	{
+		return new self($properties['value']);
```

Comment:
Since `value` is always defined by  the specific class and not in the trait, I'd more the method to each class, seems like less magic.